src/program/mod.rs

```diff
@@ -3,125 +3,91 @@
 //! event loop and render the images to screen
 
+mod render;
+pub use self::render::*;
 use crate::cli;
+use crate::paths::Paths;
+use crate::screen::Screen;
 use crate::ui::{self, Action};
 use core::cmp;
 use fs_extra::file::copy;
 use fs_extra::file::move_file;
 use fs_extra::file::remove;
-use sdl2::image::LoadTexture;
 use sdl2::rect::Rect;
-use sdl2::render::{TextureCreator, WindowCanvas};
-use sdl2::video::WindowContext;
+use sdl2::render::{Canvas, TextureCreator};
+use sdl2::ttf::Sdl2TtfContext;
+use sdl2::video::{Window, WindowContext};
 use sdl2::Sdl;
 use std::io::ErrorKind;
 use std::path::PathBuf;
 use std::time::Duration;
 
-/// Compute increment of skips
-/// Does not account for overflow or underflow of vector
-fn compute_skip_size(images: &[PathBuf]) -> usize {
-    let chunks = 10usize;
-    let skip_size: usize = (images.len() as usize / chunks) as usize + 1usize;
-
-    // Skip increment must be at least 1
-    cmp::max(1usize, skip_size)
-}
-
 /// Program contains all information needed to run the event loop and render the images to screen
-pub struct Program {
-    sdl_context: Sdl,
-    canvas: WindowCanvas,
-    texture_creator: TextureCreator<WindowContext>,
-    images: Vec<PathBuf>,
-    dest_folder: PathBuf,
-    index: usize,
+pub struct Program<'a> {
+    screen: Screen<'a>,
+    paths: Paths,
     ui_state: ui::State,
 }
 
-impl Program {
+impl<'a> Program<'a> {
     /// init scaffolds the program, by making a call to the cli module to parse the command line
     /// arguments, sets up the sdl context, creates the window, the canvas and the texture
     /// creator.
-    pub fn init() -> Result<Program, String> {
+    pub fn init(
+        ttf_context: &'a Sdl2TtfContext,
+        sdl_context: Sdl,
+        canvas: Canvas<Window>,
+        texture_creator: &'a TextureCreator<WindowContext>,
+    ) -> Result<Program<'a>, String> {
         let args = cli::cli()?;
         let images = args.files;
         let dest_folder = args.dest_folder;
-        let sdl_context = sdl2::init()?;
-        let video = sdl_context.video()?;
-        let window = video
-            .window(
-                "rust-sdl2 demo: Video",
-                video.display_bounds(0).unwrap().width(),
-                video.display_bounds(0).unwrap().height(),
-            )
-            .position_centered()
-            .resizable()
-            .build()
-            .map_err(|e| e.to_string())?;
-
-        let canvas = window
-            .into_canvas()
-            .software()
-            .build()
-            .map_err(|e| e.to_string())?;
-        let texture_creator = canvas.texture_creator();
-        let ui_state = ui::State {
-            left_shift: false,
-            right_shift: false,
+        let glob = args.search;
+
+        let current_dir = match std::env::current_dir() {
+            Ok(c) => c,
+            Err(_) => PathBuf::new(),
+        };
+        let font_path = PathBuf::from("/usr/local/share/Roboto-Medium.ttf");
```

Comment:
You're quite right. I was getting very frustrated with SDL's insistence on loading the font from a path. But actually, on second look this might allow me to load from bytes, thus getting rid of the makefile entirely. I'll give it a try later. 

https://rust-sdl2.github.io/rust-sdl2/sdl2/rwops/struct.RWops.html

Reply:
OK, that was a lot easier than expected.